test-harness/infratests/infratests.go

```diff
@@ -0,0 +1,169 @@
+/*
+Package `infratests` is intended to act as a testing harness that makes testing Terraform templates
+easy and efficient. The goal of this package is to minimize the boiler plate code required to effectively
+test terraform implementations.
+
+The current implementation is focused only on unit tests but it will be expanded to harness integration tests
+as well.
+*/
+package infratests
+
+import (
+	"errors"
+	"fmt"
+	"os"
+	"path"
+	"testing"
+
+	"github.com/gruntwork-io/terratest/modules/random"
+	"github.com/gruntwork-io/terratest/modules/terraform"
+	terraformCore "github.com/hashicorp/terraform/terraform"
+)
+
+type ResourceAttributeValueMapping map[string]map[string]string
+type TerraformPlanValidation func(goTest *testing.T, plan *terraformCore.Plan)
+
+// Holds metadata required to execute a unit test against a test against a terraform template
+type UnitTestFixture struct {
+	GoTest                *testing.T         // Go test harness
+	TfOptions             *terraform.Options // Terraform options
+	ExpectedResourceCount int                // Expected # of resources that Terraform should create
+	// map of maps specifying resource <--> attribute <--> attribute value mappings
+	ExpectedResourceAttributeValues ResourceAttributeValueMapping
+	PlanAssertions                  []TerraformPlanValidation // user-defined plan assertions
+}
+
+// Executes terraform lifecycle events and verifies the correctness of the resulting terraform.
+// The following actions are coordinated:
+//	- Run `terraform init`
+//	- Create new terraform workspace. This helps prevent accidentally deleting resources
+//	- Run `terraform plan`
+//	- Validate terraform plan file.
+func RunUnitTests(fixture *UnitTestFixture) {
+	terraform.Init(fixture.GoTest, fixture.TfOptions)
+
+	workspace_name := random.UniqueId()
+	terraform.WorkspaceSelectOrNew(fixture.GoTest, fixture.TfOptions, workspace_name)
+	defer terraform.RunTerraformCommand(fixture.GoTest, fixture.TfOptions, "workspace", "delete", workspace_name)
+	defer terraform.WorkspaceSelectOrNew(fixture.GoTest, fixture.TfOptions, "default")
+
+	tf_plan_file_path := "out.plan"
+	terraform.RunTerraformCommand(
+		fixture.GoTest,
+		fixture.TfOptions,
+		terraform.FormatArgs(fixture.TfOptions, "plan", "-input=false", "-out", tf_plan_file_path)...)
+	defer os.Remove(tf_plan_file_path)
```

Comment:
Do you foresee potential naming conflicts with multiple unit test classes for the same template considering that we're writing out a general `out.plan`?

Reply:
If the tests are palatalized then yes. Good catch, I'll make the filename random.